File: src/codec.jl
# The encoded-array codec — the whole of either side's payload knowledge.
#
# A numeric array anywhere in a payload, at any nesting depth, travels as
#   {"$wire": "f32", "shape": [3, 264], "off": 4096}
# so a decoder never needs the payload's schema, and nothing else about a payload's structure is
# interpreted anywhere. The bytes themselves sit in the frame's region — see `Frame` below.
#
# `shape` is row-major, the last dimension varying fastest, which is the reverse of Julia's
# column-major `size`. Stated that way the flat byte order is the same on both sides, so a browser
# typed array and the Julia array it came from agree on element order without a permutation. A flat
# array states `shape: [N]`.
#
# The dtype, the shape and the bytes stay three separate fields, so a reader takes the bytes as they
# are and never copies them to interpret them.
#
# Five dtypes travel. An element type with no tag of its own is converted to the tag that carries it
# without loss, and refused when no such tag exists — never narrowed, never wrapped around:
#
# | Julia eltype              | travels as | |
# |---------------------------|------------|---------------------------------------------------|
# | `Float32` `Float64`       | f32 f64    | |
# | `UInt8` `UInt32` `Int32`  | u8 u32 i32 | |
# | `Bool`                    | u8         | JS has no boolean typed array                     |
# | `Int8` `Int16` `Int64`    | i32        | every value must fit `Int32`, or it is an error   |
# | `UInt16` `UInt64`         | u32        | every value must fit `UInt32`, or it is an error  |
# | `Float16`                 | f32        | always exact                                      |
# | any other `<: Number`     | error      | `Complex`, `Rational`, `Int128` and the like      |
# | anything not `<: Number`  | JSON list  | strings, nested payload objects, `Array{Any}`     |

# One message and the array bytes behind it, before either is laid out for a transport.
#
# The builders in `messages.jl` return this rather than packed bytes, so a recording can store a
# readable header and `send_message` can still take hand-written JSON. `pack` is called only where a
# frame reaches a transport.

"""
    Frame(header, blobs = UInt8[])

One wire message: `header` is the JSON-RPC text, `blobs` is the region its encoded arrays point
into. A message with no arrays has an empty region.
"""
struct Frame
    header::String
    blobs::Vector{UInt8}
end
Frame(header::AbstractString) = Frame(String(header), UInt8[])

"""
    pack(f::Frame) -> Vector{UInt8}

The bytes one WebSocket binary message carries:

    [u32 headerLen][header][pad to 8][region]

All integers are little-endian, which every target platform is. The pad puts the region on a
multiple of 8, so a reader can take a `Float64Array` view over any array in it.
"""
function pack(f::Frame)
    h = codeunits(f.header)
    io = IOBuffer(; sizehint = 4 + length(h) + 8 + length(f.blobs))
    write(io, UInt32(length(h)))
    write(io, h)
    write(io, zeros(UInt8, region_pad(4 + length(h))))
    write(io, f.blobs)
    return take!(io)
end

"""
    unpack(bytes) -> Frame

The frame [`pack`](@ref) laid out, split back into its header and its region. Reads a frame captured
off the socket, and is what `replay` and the test clients use.
"""
function unpack(bytes::AbstractVector{UInt8})
    length(bytes) ≥ 4 || error("wire: a frame is at least 4 bytes; got $(length(bytes))")
    n = Int(only(reinterpret(UInt32, bytes[1:4])))
    4 + n ≤ length(bytes) ||
        error("wire: a header of $n bytes runs past a frame of $(length(bytes))")
    # A truncated frame leaves an empty region rather than an out-of-range index; the bounds check
    # in `decode_array` is what refuses the arrays that then point past it.
    start = min(4 + n + region_pad(4 + n), length(bytes))
    return Frame(String(bytes[5:(4 + n)]), bytes[(start + 1):end])
end

# Zero bytes to write after `n` bytes to reach the next multiple of 8.
region_pad(n::Integer) = (-n) & 7

const WIRE_TAG = Dict{DataType,String}(
    Float32 => "f32", Float64 => "f64", UInt8 => "u8", UInt32 => "u32", Int32 => "i32",
)
const WIRE_TYPE = Dict(tag => T for (T, tag) in WIRE_TAG)

# Element types carried by a tag other than their own, through a lossless conversion.
const WIRE_CONVERT = Dict{DataType,DataType}(
    Bool => UInt8,
    Int8 => Int32, Int16 => Int32, Int64 => Int32,
    UInt16 => UInt32, UInt64 => UInt32,
    Float16 => Float32,
)

"""
    encode_arrays(payload, region::IO)

Return `payload` with every numeric array inside it, at any depth, replaced by its encoded form, and
write the bytes of each into `region`. Dicts, named tuples, tuples and non-numeric arrays are walked;
everything else passes through. The caller takes the region and the returned value as one frame.

Every array is padded to an offset that is a multiple of 8, whatever its dtype: one rule instead of
a table, and a reader can then take a `Float64Array` view over any of them. The wasted bytes are at
most 7 per array.

**Build a mixed payload with a container that holds its element types apart.** A Julia array literal
promotes, and `promote_type(Vector{Float64}, Vector{UInt8})` is `Vector{Float64}` — so
`["a" => f64s, "b" => u8s]` silently converts the `UInt8` array to `Float64`. That is eight times
the bytes and the wrong dtype on the wire, with no error anywhere. Write `(; a = f64s, b = u8s)`, or
annotate the literal as `Pair{String,Any}[...]`.

A vector of structs whose fields share one type travels as `reinterpret(reshape, S, v)` with no
codec change — a bare `Vector{ECEF}` is not `<: Number` and would otherwise become JSON objects.

An array whose element type is not one of the five wire dtypes is converted to the dtype that
carries it without loss, and an array no dtype can carry is an error rather than a silent JSON list:

| Julia eltype | travels as | |
|---|---|---|
| `Float32`, `Float64` | `f32`, `f64` | |
| `UInt8`, `UInt32`, `Int32` | `u8`, `u32`, `i32` | |
| `Bool` | `u8` | JS has no boolean typed array; read a flag as `data[i] !== 0` |
| `Int8`, `Int16`, `Int64` | `i32` | error unless every value fits `Int32` |
| `UInt16`, `UInt64` | `u32` | error unless every value fits `UInt32` |
| `Float16` | `f32` | always exact |
| any other `<: Number` | error | `Complex`, `Rational`, `Int128` and the like |
| anything not `<: Number` | JSON list | strings, nested payload objects, `Array{Any}` |

An integer too large for its dtype is refused, never wrapped: convert it to `Float64`, which is
exact to `2^53`, or split the value.
"""
encode_arrays(x, region::IO) = x
encode_arrays(d::AbstractDict, region::IO) =
    Dict{String,Any}(string(k) => encode_arrays(v, region) for (k, v) in d)
encode_arrays(t::NamedTuple, region::IO) =
    Dict{String,Any}(string(k) => encode_arrays(v, region) for (k, v) in pairs(t))
encode_arrays(t::Tuple, region::IO) = Any[encode_arrays(v, region) for v in t]

function encode_arrays(a::AbstractArray, region::IO)
    tag = get(WIRE_TAG, eltype(a), nothing)
    if tag === nothing
        # An array of anything but numbers is a payload structure, not data: walk it element by
        # element, which is what carries strings and nested payload objects.
        eltype(a) <: Number || return Any[encode_arrays(v, region) for v in a]
        return encode_arrays(to_wire_eltype(a), region)
    end
    # Julia is little-endian on the target platforms, so a dense array's bytes are the wire bytes.
    flat = a isa Array ? vec(a) : vec(collect(a))
    write(region, zeros(UInt8, region_pad(position(region))))
    off = position(region)
    write(region, reinterpret(UInt8, flat))
    return Dict{String,Any}("\$wire" => tag, "shape" => reverse(collect(size(a))), "off" => off)
end

# The array `a` with the element type that carries it on the wire, so a conversion can only ever be
# exact. The range is checked ahead of the conversion for the message: the `InexactError` the
# conversion itself would raise names neither the dtype involved nor a way out.
function to_wire_eltype(a::AbstractArray)
    T = eltype(a)
    S = get(WIRE_CONVERT, T, nothing)
    S === nothing && error(
        "wire: no dtype carries $T arrays. Arrays travel as f32 f64 u8 u32 i32, and Bool, Int8, " *
        "Int16, Int64, UInt16, UInt64 and Float16 convert into one of those.")
    if S <: Integer && !isempty(a)
        lo, hi = extrema(a)
        typemin(S) <= lo && hi <= typemax(S) || error(
            "wire: $lo … $hi does not fit $S, which is what $T arrays travel as. Convert to " *
            "Float64, which is exact for |v| ≤ 2^53, or split the value.")
    end
    return Array{S}(a)
end

"""
    decode_arrays(payload, region = UInt8[])

Return `payload` with every encoded array inside it, at any depth, replaced by the array it carries,
reading the bytes out of `region`. Everything else — including a value that merely carries a `\$wire`
key — passes through.

An array is recognised by its `off`, which is where in `region` its bytes start.
"""
decode_arrays(x, region = UInt8[]) = x
decode_arrays(v::AbstractVector, region = UInt8[]) = Any[decode_arrays(x, region) for x in v]

function decode_arrays(d::AbstractDict, region = UInt8[])
    is_wire_array(d) && return decode_array(d, region)
    return Dict{String,Any}(string(k) => decode_arrays(v, region) for (k, v) in d)
end

# Anything but a known dtype tag alongside an offset into the region is an ordinary payload value
# that happens to have a `$wire` key.
function is_wire_array(d::AbstractDict)
    tag = get(d, "\$wire", nothing)
    tag isa AbstractString && haskey(WIRE_TYPE, tag) || return false
    return get(d, "off", nothing) isa Integer
end

function decode_array(d::AbstractDict, region = UInt8[])
    tag = d["\$wire"]
    T = WIRE_TYPE[tag]
    shape = get(d, "shape", nothing)
    # The length is derived rather than carried, so this bounds check is the whole of what stands
    # between a malformed frame and a read past the region.
    shape === nothing && error("wire: an array in a region must state its shape")
    dims = reverse(Int.(shape))
    off, n = Int(d["off"]), prod(dims) * sizeof(T)
    0 ≤ off && off + n ≤ length(region) || error(
        "wire: a $tag array of $n bytes at offset $off runs past a region of " *
        "$(length(region)) bytes")
    return reshape(collect(reinterpret(T, region[(off + 1):(off + n)])), dims...)
end

# The other thing that flips at the wire boundary, beside the row-major shape above: an index.

"""
    to_wire_index(i) -> Integer

A 1-based Julia index as the 0-based index the wire carries. Its inverse is
[`from_wire_index`](@ref). Broadcast it over an array of indices.
"""
to_wire_index(i::Integer) = i - one(i)

"""
    from_wire_index(i) -> Integer

A 0-based wire index as the 1-based index Julia carries. Its inverse is [`to_wire_index`](@ref).
"""
from_wire_index(i::Integer) = i + one(i)

File: lib/core/src/codec.ts
// The encoded-array codec — the whole of the Core's payload knowledge.
//
// A numeric array anywhere in any payload, at any nesting depth, travels as
//   { $wire: "f32", shape: [3, 264], off: 4096 }
// so a decoder needs no schema. Its bytes sit in the frame's region, at `off` bytes from the start
// of that region. `shape` is row-major, the last dimension varying fastest; Julia, being
// column-major, reverses it at its own boundary.
//
// Typed arrays are host-endian, and every target this runs on is little-endian.
//
// One value here encodes: a canvas capture, which is the only array that travels upward
// (ADR-0033). It rides the same `u8` tag as everything else, so the upward half of the protocol
// uses this contract and does not extend it. See `docs/protocol.md`.

const CTORS = {
  f32: Float32Array,
  f64: Float64Array,
  u8: Uint8Array,
  u32: Uint32Array,
  i32: Int32Array,
} as const;

export type Dtype = keyof typeof CTORS;
type TypedArray = InstanceType<(typeof CTORS)[Dtype]>;

/** An array on the wire: its bytes are `off` bytes into the frame's region. */
export interface WireArray {
  $wire: Dtype;
  shape: number[];
  off: number;
}

/** An array in a payload, decoded. */
export interface NdArray {
  data: TypedArray;
  shape: number[];
}

/** The values one keyframe reads, as an offset into the array that carries them. */
export interface Block {
  readonly data: ArrayLike<number>;
  readonly offset: number;
  /** Values in the block: the product of the trailing dimensions. */
  readonly len: number;
  /** True when the values switch at keyframe crossings. */
  readonly keyframed: boolean;
}

/**
 * The block keyframe `k` addresses, or null where the array says nothing about `k`.
 *
 * The caller names the **base rank** of the form it expects — 1 for a value per entity, 3 for a
 * `[H, W, 4]` raster. An array at or below that rank holds one value for the whole window, and every
 * keyframe reads all of it. An array one rank above it carries a leading keyframe axis, so keyframe
 * `k` is the contiguous block at `k × len`, where `len` is the product of the trailing dimensions.
 *
 * `count` is the keyframes the window carries. An array whose leading axis disagrees with it throws:
 * a payload that claims seven keyframes inside a window of five is a bug worth being loud about,
 * rather than a short block read as data.
 */
export function blockAt(a: NdArray, k: number, baseRank: number, count: number): Block | null {
  const { data, shape } = a;
  if (shape.length <= baseRank) {
    return { data, offset: 0, len: data.length, keyframed: false };
  }
  if (shape.length !== baseRank + 1) {
    throw new Error(`codec: shape [${shape}] is more than one rank above base rank ${baseRank}`);
  }
  if (shape[0] !== count) {
    throw new Error(`codec: shape [${shape}] has ${shape[0]} keyframes, the window carries ${count}`);
  }
  if (k < 0 || k >= count) return null;
  // Multiplied out in place: a module on the interpolation path calls this every tick, and the
  // trailing dimensions are not worth an array of their own.
  let len = 1;
  for (let axis = 1; axis < shape.length; axis++) len *= shape[axis];
  return { data, offset: k * len, len, keyframed: true };
}

/**
 * Replace every encoded array inside `value` with its `{data, shape}` form, reading the bytes out
 * of `region`.
 *
 * Each array is a **view into `region`**, not a copy: copying costs about 1 ms per megabyte, which
 * is most of what the region exists to save. A module that keeps a slice past its window keeps the
 * whole frame alive with it — `new Float32Array(a.data)` detaches a copy. See `docs/module-api.md`.
 */
export function decodeArrays(value: unknown, region: Uint8Array): unknown {
  if (Array.isArray(value)) return value.map((v) => decodeArrays(v, region));
  if (value === null || typeof value !== "object") return value;
  if (isWireArray(value)) return decodeArray(value, region);
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value)) out[k] = decodeArrays(v, region);
  return out;
}

// An object carrying anything but a known dtype tag, a shape and an offset is an ordinary payload
// value that happens to have a `$wire` key, and passes through untouched.
function isWireArray(v: object): v is WireArray {
  const o = v as Partial<WireArray>;
  return (
    typeof o.$wire === "string" &&
    o.$wire in CTORS &&
    typeof o.off === "number" &&
    Array.isArray(o.shape)
  );
}

/**
 * Whether a decoded value is one of the typed arrays `decodeArrays` produces. Exported because
 * every consumer of a payload has to ask it, and the answer belongs next to the shape it describes.
 */
export function isNdArray(v: unknown): v is NdArray {
  const o = v as Partial<NdArray> | null;
  return !!o && typeof o === "object" && ArrayBuffer.isView(o.data) && Array.isArray(o.shape);
}

/**
 * Read a declared list of numbers as a plain list. A short list travels as JSON, a long list travels
 * as an encoded array, and a field the declaration omits gives the empty list.
 */
export const numbers = (v: number[] | NdArray | undefined): number[] =>
  v === undefined ? [] : isNdArray(v) ? Array.from(v.data) : v;

function decodeArray(w: WireArray, region: Uint8Array): NdArray {
  const Ctor = CTORS[w.$wire];
  const count = elementCount(w.shape);
  const len = count * Ctor.BYTES_PER_ELEMENT;
  // The length is derived from the shape rather than carried, so this is the whole of what stands
  // between a malformed frame and a view over bytes that never arrived.
  if (w.off < 0 || w.off + len > region.byteLength) {
    throw new Error(
      `codec: a ${w.$wire} array of ${len} bytes at offset ${w.off} runs past a region of ` +
        `${region.byteLength} bytes`,
    );
  }
  // Every offset is a multiple of 8, so a Float64Array view is always alignment-legal. An offset
  // that is not means the frame is wrong, and the constructor throwing says so.
  const buf = region.buffer as ArrayBuffer;
  return { data: new Ctor(buf, region.byteOffset + w.off, count), shape: w.shape };
}

/**
 * Describe `bytes` as the array a payload points at.
 *
 * The caller sends `bytes` as the whole region of its frame, so the offset is 0 and the shape is
 * the byte count. A canvas capture is the one array that travels upward, and it needs one array
 * per region (ADR-0033).
 */
export function encodeU8(bytes: Uint8Array): WireArray {
  return { $wire: "u8", shape: [bytes.length], off: 0 };
}

const elementCount = (shape: number[]): number => shape.reduce((a, b) => a * b, 1);
